deps/thpool/thpool.c

```diff
@@ -346,25 +456,186 @@ void redisearch_thpool_destroy(redisearch_thpool_t* thpool_p) {
     thread_destroy(thpool_p->threads[n]);
   }
   rm_free(thpool_p->threads);
+  rm_free(thpool_p->name);
   rm_free(thpool_p);
 }
 
 /* Pause all threads in threadpool */
 void redisearch_thpool_pause(redisearch_thpool_t* thpool_p) {
+  if (!thpool_p) return;
+
+  while (thpool_p->flags & RS_THPOOL_F_RESUME) {
+    sleep(1);
+    RedisModule_Log(
+        NULL, "warning",
+        "redisearch_thpool_pause(): waiting for %s threadpool to finish resuming process.",
+        thpool_p->name);
+  }
   size_t n;
-  for (n = 0; n < thpool_p->num_threads_alive; n++) {
-    pthread_kill(thpool_p->threads[n]->pthread, SIGUSR2);
+	pthread_t caller = pthread_self();
+
+  // save the current thpool to the global pointer
+  g_curr_threadpool = thpool_p;
+
+  size_t threadpool_size = thpool_p->num_threads_alive;
+
+  int called_by_threadpool = 0;
+
+  // set hold flag
+  thpool_p->flags |= RS_THPOOL_F_PAUSE;
+
+  // zero threads' ids. This number will be increased atomically by each
+  // thread so it can write to its own unique idx in the output array.
+  g_threads_ids = 0;
+
+	for(n = 0; n < threadpool_size; n++) {
+		// do not pause caller
+		if(thpool_p->threads[n]->pthread != caller) {
+			pthread_kill(thpool_p->threads[n]->pthread, SIGUSR2);
+		} else {
+      // The calling thread belongs to this thread pool
+      called_by_threadpool = 1;
+      thpool_p->flags |= RS_THPOOL_F_CONTAINS_HANDLING_THREAD;
+    }
+	}  
+
+  if (threadpool_size) {
+    // wait until all the threads in the thpool are paused (except for the caller)
+    size_t expected_new_paused_count = threadpool_size - called_by_threadpool;
+    clock_t start = clock();
+    size_t paused_threads = g_threads_ids;
+    while(g_threads_ids < expected_new_paused_count) {
```

Comment:
No need for atomicity with `g_threads_ids`?

Reply:
It can only grow (up to the number of threads in the threadpool) so in the worst case we execute another loop

Reply:
On 32bit arch it could be much worse

Reply:
explain?

Reply:
If writing is done in 2 separate operations (2 32-bit registers) and context switch occurs in the between, the value may be completely wrong

Reply:
only the read operation is not atomic

Reply:
Oh, you mean writing to paused_threads?

Reply:
Changed g_threads_paused_cnt to atomic